### PR TITLE
Add EngineData container to attach external data to ENGINE

### PR DESCRIPTION
**Pull request by ZacharyEspiritu**
Part of #12. Blocked by #59.

Adds `EngineData` class for holding all of the external data that is attached to a given `ENGINE`; to be used in conjunction with the ex_data utilities.

### CODE REVIEW

**Comment by ZacharyEspiritu** [comment]
I've marked this as ready for review because even though it technically depends on #59, it appears that GitHub will, at least, automatically update the base branch of the PR from `rsa-key` to `master` once #59 is merged so then this can be merged immediately after. (I don't know if this automatic base branch change will automatically dismiss reviews, which might defeat the purpose. But this might be something to try.)

**Review by tonyliaoss — DISMISSED** [review]

File: src/bridge/ex_data_util/engine_data.h
Comment (ajfeldma-google) [inline src/bridge/ex_data_util/engine_data.h]: I don't think `inline` is needed on methods

File: src/backing/client/client.h
Comment (ajfeldma-google) [inline src/backing/client/client.h:40]: Are changes to this file meant to be part of this PR?

**Review by ajfeldma-google — CHANGES_REQUESTED** [review]

File: src/backing/client/client.h
Comment (ZacharyEspiritu) [reply src/backing/client/client.h:40]: Yes---added to allow for EngineData to return a const reference to Client and then callers to call methods on that client:

```
engine_data()->client().GetPublicKey
```

**Review by ajfeldma-google — APPROVED** [review]

**Review by tonyliaoss — APPROVED** [review]

**Pull request merged**
